OEP-6/OEP-6.mediawiki

```diff
@@ -0,0 +1,452 @@
+
+===Asynchronicity and error handling===
+All the functions except '''Utils''' component are communicating with extension through asynchronious message channel. Therefore all the methods are returning Promises. 
+
+The promises will be either resolved immediately if no interaction with user UI or blockchain is required or later when the user action takes place/blockchain responds. In case the call to method trigger an error, the error code is transmitted in rejected promise. Specific error codes are described with every method.
+
+===Account/Identity management===
+Because dAPI shifts the issue of trust from dApp to dApp provider, all account and identity management is external to the dApp. Therefore there are no methods which directly handle private keys. The dApp won't need to sign the transaction itself. 
+
+Any time dApp makes a call that requires a private key to sign something (makeTransfer, sign), dApp provider will inform user about the action and prompt him for permission.
+
+dApp provider can even forward the request to external signing mechanism as part of the process. This way the dApp does not need to specifically integrate with such mechanism.
+
+===Encoding===
+The interaction with dAPI often requires specification of '''Address''', '''Identity''', '''Contract address''' or '''Public Key'''. Those concepts are represented by string representation with specific encodings.
+
+====Address====
+* 34 bytes long
+* using Base58 check encoding
+* starts with A letter
+
+====Contract address====
+* 40 bytes long
+* using Hex encoding
+
+====Identity====
+* 42 bytes long
+* using Base58 check encoding for 34 byte suffix
+* starts with 'did:ont:' prefix
+
+====Public Key====
+* 33/35 bytes long
+* using Hex encoding
+
+===Complex structures===
+API specification is a complex document. Every method has some inputs and outputs. Although we use the primitive types (numbers, strings, booleans) whenever possible, there are places where a complex object is required. To precisely describe the structure of those objects, we'd chosen Typescript syntax.
+
+===Components===
+Although this proposal is bringing clear and simple API for the dApps, the individual functions can be divided into these components:
+
+* '''Network''', a thin wrapper around the Ontology Node API, masking the complexity of rpc/rest calls and web-sockets with Request-Response facade.
+
+* '''Provider''', functions for getting information about dAPI provider
+
+* '''Asset''', functions for transferring assets between user account and others.
+
+* '''Identity''', functions for interacting with own ONT-ID identity.
+
+* '''SmartContract''', a high level wrapper around the Smart Contract invocation and deployment.
+
+* '''Message''', functions for signing arbitrary messages.
+
+* '''Utils''', a group of utility function for encoding and decoding the data from/to blockchain.
+
+===Network===
+A network API consists of:
+
+<pre>
+type Network = 'MAIN' | 'TEST' | 'PRIVATE';
```

Comment:
For now, the getNetwork() method will provide only the information that wallet is not connected to MAIN-NET or TEST-NET. I might add method getNetworkAddress() which will return the specific url of the network (in case of main/test net the corresponding address of the network = used Node address to be precise)

Reply:
I refactored it a little different. The getNetwork now returns an object with type and address property.